text/0026-semantic-versioning.md

```diff
@@ -0,0 +1,80 @@
+
+The [CNB Buildpacks spec](https://github.com/buildpacks/spec/blob/buildpack/v0.6/buildpack.md) says:
+
+> The buildpack version:
+>
+> - MUST be in the form <X>.<Y>.<Z> where X, Y, and Z are non-negative integers and must not contain leading zeros.
+>    - Each element MUST increase numerically.
+>    - Buildpack authors will define what changes will increment X, Y, and Z.
+
+
+It does not define what changes will increment the Major (`X`), Minor (`Y`) and Patch (`Z`) versions
+of buildpack releases.
+
+This RFC attempts to set guidelines and good practices for Paketo Buildpack authors/maintainers to follow
+when a buildpack release has to be created.
+
+## Motivation
+
+As a Buildpack/Builder maintainer,<br/>
+I need a release versioning guideline,
+so I can make my release versions consistent.
+
+As a Buildpack/Builder user,<br/>
+I want to know how versioning is done,
+so I can use the appropriate version and
+decide when to upgrade.
+
+## Semantics
+
+Semantic Versioning 2.0.0 defines versioning syntax as follows [semver.org](https://semver.org/):
+
+> Given a version number MAJOR.MINOR.PATCH, increment the:
+> 
+>   - MAJOR version when you make incompatible API changes,
+>   - MINOR version when you add functionality in a backwards compatible manner, and
+>   - PATCH version when you make backwards compatible bug fixes.
+
+The proposal is to conform to the above semantics as much as possible.
+
+### Changes
+
+In terms of Paketo Buildpacks/Builders, this translates to:
+
+* Bumping MAJOR is analogous to an "`app or related buildpacks may not work without change`" relation between releases.
+* Bumping MINOR is analogous to an "`updates or adds features`" relation between releases.
+* Bumping PATCH is analogous to a "`fixes issue`" relation between releases.
+
+The following table lists the usual circumstances/events where bumping a particular component is expected:
+
+#### Buildpack
+| `Major` | `Minor` | `Patch` |
+|-|-|-|
+| \<Any incompatible change\> | \<Feature additions but compatible with previous release\> | \<Fixes but no major feature changes\> |
+| Change in Buildpack ID | Addition/removal of dependency<br>(but default-version not changed to a version with breaking change) | Bug fixes |
+| Change in supported Stacks | Buildpack runs a new command the causes change in behavior<br>(e.g. x-install buildpack adds running "x clean") | Performance improvement, internal tweaks |
+| Change in API (Provides/Requires) | Change of required build tool version to package buildpack<br>(e.g. go version) | CVE fixes |
+| Default version of dependency bumped to incompatible version<br>(e.g. php7 to php8) | New configuration option (e.g. new environment variable support) |  |
+| Removal of configuration options<br>(e.g. buildpack.yml) | Change in Buildpack API version |  |
+| Removal of publishing to a certain image registry<br>(e.g. gcr.io, dockerhub) | Publishing to a new image registry |  |
```

Comment:
I'm not sure I follow why we'd need to change the version of the buildpack given where it is published. There shouldn't be anything in the image that indicates that it can from any registry in particular. To me, no longer updating the images at a certain registry location doesn't strike me as a breaking change. It certainly isn't breaking to those using the image at the registries that still host the latest images, and it isn't breaking for those using the older images on the registry that will no longer be updated.

Reply:
I added this on purpose as an exception to the breaking-change rule that could warrant a major bump. It could be better changed to `Change in official image URI`, the proposal here being - since consumption of buildpack will largely be driven through official image registry uri and not via .cnb downloads, consumers will have to make changes to their build systems to keep consuming the same buildpack.

Reply:
But that change isn't to the buildpack. The buildpack doesn't contain breaking changes. If you are already using the new registry, seeing a major version bump, for apparently no reason, would be really surprising. I agree with the intent to ensure folks get ample notification that the buildpack will be moving registry locations, but we shouldn't try to convey that through a version number.

Reply:
Sounds good. Removed the entry.